lib/puppet/type/nssdb.rb

```diff
@@ -0,0 +1,69 @@
+Puppet::Type.newtype(:nssdb) do
+  desc "Generates an empty NSS database"
+
+  ensurable
+
+  newparam(:nssdb_dir, :namevar => true) do
+    desc "Path to NSS database directory"
+    isrequired
+  end
+
+  newparam(:password_file) do
+    desc "Path to file containing the NSS database password"
+    isrequired
+  end
+
+  newparam(:owner, parent: Puppet::Type::File::Owner) do
+    desc "Specifies the owner of the NSS database directory and files. Valid options: a string containing a username or integer containing a uid."
+  end
+
+  newparam(:group, parent: Puppet::Type::File::Group) do
+    desc "Specifies a permissions group for the NSS database directory and files. Valid options: a string containing a group name or integer containing a gid."
+  end
+
+  newparam(:mode, parent: Puppet::Type::File::Mode) do
+    desc "Specifies the permissions mode of the NSS database files. Valid options: a string containing a permission mode value in octal notation."
+  end
+
+  newparam(:password_mode, parent: Puppet::Type::File::Mode) do
+    desc "Specifies the permissions mode of the NSS database password file. Valid options: a string containing a permission mode value in octal notation."
+  end
+
+  autorequire(:file) do
+    [self[:password_file], self[:nssdb_dir]]
+  end
+
+  def generate
+    file_opts = {}
+
+    [:owner, :group].each do |param|
+      file_opts[param] = self[param] unless self[param].nil?
+    end
+
+    excluded_metaparams = [:before, :notify, :require, :subscribe, :tag]
+
+    Puppet::Type.metaparams.each do |metaparam|
+      unless self[metaparam].nil? || excluded_metaparams.include?(metaparam)
+        file_opts[metaparam] = self[metaparam]
+      end
+    end
+
+    resources = []
+    ensure_condition = self[:ensure] == :absent
+
+    dir_opts = file_opts.dup
+    dir_opts.merge({ :path => resource[:nssdb_dir] })
+    dir_opts.merge({ :mode => resource[:mode] }) unless self[:mode].nil?
+    dir_opts.merge({ :ensure => ensure_condition ? :absent : :directory })
+    dir_opts.merge({ :recurse => true })
```

Comment:
This is what I landed on for now to manage attributes for all generated files, but I forsee a few problems:

1. If the password_file exists without the nssdb_dir, this should cause a duplicate resource declaration error

2. I'm unable to manage the directory mode ('0755' in the previous implementation) separately from the file mode (0640). In particular, how is it even possible to set execute permission on the directory without setting it on all files within, when using `recurse => true` ?

A possible solution to these problems could be to use: https://forge.puppet.com/modules/npwalker/recursive_file_permissions

Your feedback and suggestions would be appreciated @ehelms @ekohl

Reply:
Unsure at the time if this is what is causing `Error: /Stage[main]/Certs::Ssltools::Nssdb/Nssdb[/etc/pki/katello/nssdb]: Failed to generate additional resources using 'generate': Title or name must be provided`... will do some more testing

Reply:
Given that the "nss database" is thought of as a directory, perhaps we should not try to manage the directory and Files in the directory separately user/group/permission wise. We should set them all the same.

And to that end, perhaps we should not put the password file in there either and thereby treat it like we do the other resources? For example, the file tree might be:

```
/etc/qpid/certs
/etc/qpid/certs/nssdb-password_file
/etc/qpid/certs/nssdb/key3.db
/etc/qpid/certs/nssdb/cert9.db
```

Reply:
OK, latest update again removes the password file management from the provider/type, and moves the location into the parent dir.